make/catalog.mk

```diff
@@ -7,9 +7,22 @@ CATALOG_IMG ?= $(IMAGE_TAG_BASE)-catalog:$(IMAGE_TAG)
 CATALOG_FILE = $(PROJECT_PATH)/catalog/kuadrant-operator-catalog/operator.yaml
 CATALOG_DOCKERFILE = $(PROJECT_PATH)/catalog/kuadrant-operator-catalog.Dockerfile
 
+# Quay image default expiry
+QUAY_IMAGE_EXPIRY ?= never
+
+# A LABEL that can be appended to a generated Dockerfile to set the Quay image expiration through Docker arguments.
+define QUAY_DOCKERFILE_LABEL
+
+# Quay image expiry
+ARG QUAY_IMAGE_EXPIRY
+ENV QUAY_IMAGE_EXPIRY=$${QUAY_IMAGE_EXPIRY:-never}
+LABEL quay.expires-after=$${QUAY_IMAGE_EXPIRY}
+endef
+export QUAY_DOCKERFILE_LABEL
+
 $(CATALOG_DOCKERFILE): $(OPM)
 	-mkdir -p $(PROJECT_PATH)/catalog/kuadrant-operator-catalog
-	cd $(PROJECT_PATH)/catalog && $(OPM) generate dockerfile kuadrant-operator-catalog
+	cd $(PROJECT_PATH)/catalog && $(OPM) generate dockerfile kuadrant-operator-catalog -l quay.expires-after=$(QUAY_IMAGE_EXPIRY)
 catalog-dockerfile: $(CATALOG_DOCKERFILE) ## Generate catalog dockerfile.
```

Comment:
We can also allow setting the Quay image expiry by following the same pattern to how it was done for the bundle image. If we choose to implement this, the Docker label will be consistently set via build arguments. However, I'm not sure if it's worth the extra step just for the sake of consistency. If anyone has any thoughts on this, feel free to share, otherwise achieving this through `opm` labeling is simpler. 
```suggestion
	cd $(PROJECT_PATH)/catalog && $(OPM) generate dockerfile kuadrant-operator-catalog 
	echo "$$QUAY_DOCKERFILE_LABEL" >> $(CATALOG_DOCKERFILE)
catalog-dockerfile: $(CATALOG_DOCKERFILE) ## Generate catalog dockerfile.

# Build a catalog image by adding bundle images to an empty catalog using the operator package manager tool, 'opm'.
# Ref https://olm.operatorframework.io/docs/tasks/creating-a-catalog/#catalog-creation-with-raw-file-based-catalogs
.PHONY: catalog-build
catalog-build: ## Build a catalog image.
	# Build the Catalog
	docker build --build-arg QUAY_IMAGE_EXPIRY=$(QUAY_IMAGE_EXPIRY)  $(PROJECT_PATH)/catalog -f $(PROJECT_PATH)/catalog/kuadrant-operator-catalog.Dockerfile -t $(CATALOG_IMG)
```

Reply:
no hard opinion on this matter. For me what it matters is that the label should be there. 

I do have a "soft opinion", though. I would do the labeling using the `opm` command. Mainly because we do not have a committed version of the dockerfile, it is autogenerated. So, the expiry time is not going to be committed in the repo even though it is hardcoded in the (ephemeral) dockerfile.

Reply:
I also do not like 

```
echo "$$QUAY_DOCKERFILE_LABEL" >> $(CATALOG_DOCKERFILE)
```
if we need to do that, we do that. But modifying dynamically the dockerfile is somewhat unwanted. Can we ensure the value of `$QUAY_DOCKERFILE_LABEL` is syntactically correct?

Reply:
> if we need to do that, we do that. But modifying dynamically the dockerfile is somewhat unwanted. 

@eguzki  Yes, we don't need to do this for the **catalog** dockerfile 👍  But I think we need to do this for the **bundle** Dockerfile since this is also autogenerated, is overwritten every time and committed. 
https://github.com/Kuadrant/kuadrant-operator/blob/107947ed5b7478c523c13cb66e373394db5167f8/Makefile#L393
https://github.com/Kuadrant/kuadrant-operator/blob/107947ed5b7478c523c13cb66e373394db5167f8/bundle.Dockerfile#L23-L26

I'm not aware of a way to add extra labels to the generated bundle dockerfile via operator-sdk 🤔  

> Can we ensure the value of $QUAY_DOCKERFILE_LABEL is syntactically correct?

It should be correct, but not sure how we can ensure this? It's functioning at least since the label was set correctly for the bundle image built 🤔

Reply:
right, I ran into the same issue when wanted to add openshift specific labels to the bundle image. 

I am afraid, I do not know how to do this cleanly, either. 

We decided to have git commited the bundle.Dockerfile (and `bundle` dir) for a reason. Reason being having traceability. Updating the file on-the-fly sort of defeats that purpose. 

I propose to do what you suggest. A sacrifice for the greater good. Do no git update `bundle.Dockerfile`, change it on the fly instead.

Reply:
> I propose to do what you suggest. A sacrifice for the greater good. Do no git update bundle.Dockerfile and commit it, change it on the fly instead and revert changes when the image build is done.

@eguzki  I'm thinking of reverting the [last commit](https://github.com/Kuadrant/kuadrant-operator/pull/851/commits/e4f3ac3208d10223839c0a074987ce0b7f88c28b) for this. Reason is the CI build for the **bundle** image needs to be updated also to account for this ([why it's not working on the CI build currently](https://quay.io/repository/kuadrant/kuadrant-operator-bundle?tab=tags)) since it does not use the makefile commands to build the image, and likely be a duplicated effort to also do this in the CI.
https://github.com/Kuadrant/kuadrant-operator/blob/e4f3ac3208d10223839c0a074987ce0b7f88c28b/.github/workflows/build-images-base.yaml#L138-L168

This, to me, is not as clean as the previous approach with just always appending the neccessay `QUAY_DOCKERFILE_LABEL` in the `make bundle` command after bundle image dockerfile has been generated, which will be the actual dockerfile used when building the bundle image rather than hiding this detail 🤔 
Since we commit this Dockerfile, it's apparent from the Dockerfile that it can accept a build argument to allow expirying quay images if passed. Otherwise this detail is hidden and harder to debug in general I feel if an issue occurs.

What do you think?

Reply:
So, when you run `bundle` makefile target, `$(OPERATOR_SDK) generate bundle ... ` will overwrite the bundle.Dockerfile and then we add the label again, right? something like

```
bundle:
    .... 
    $(OPERATOR_SDK) generate bundle -q --overwrite
    .....
    @echo "$$QUAY_DOCKERFILE_LABEL" >> bundle.Dockerfile
```

If this is what you suggest, I am ok with it. And I propose:

`s/QUAY_DOCKERFILE_LABEL/QUAY_EXPIRY_TIME_LABEL/`

Or something similar.

Reply:
> So, when you run `bundle` makefile target, `$(OPERATOR_SDK) generate bundle ... ` will overwrite the bundle.Dockerfile and then we add the label again, right? If this is what you suggest, I am ok with it.

Yes, exactly 👍 

> . And I propose: `s/QUAY_DOCKERFILE_LABEL/QUAY_EXPIRY_TIME_LABEL/`. Or something similar.

Sounds good, updated to `QUAY_EXPIRY_TIME_LABEL` 👍